source/darwin/securityframework_rsa.c

```diff
@@ -0,0 +1,543 @@
+    }
+
+    SecKeyAlgorithm alg;
+    if (s_map_rsa_signing_algo_to_sec(algorithm, &alg)) {
+        return AWS_OP_ERR;
+    }
+
+    if (!SecKeyIsAlgorithmSupported(key_pair_impl->priv_key_ref, kSecKeyOperationTypeSign, alg)) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Algo is not supported for this operation");
+        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    }
+
+    CFDataRef digest_ref = CFDataCreateWithBytesNoCopy(key_pair_impl->cf_allocator, digest.ptr, digest.len, kCFAllocatorNull);
+    AWS_FATAL_ASSERT(digest_ref);
+
+    CFErrorRef error = NULL;
+    CFDataRef signature_ref = SecKeyCreateSignature(key_pair_impl->priv_key_ref, alg, digest_ref, &error);
+    if (s_reinterpret_sec_error_as_crt(error, "SecKeyCreateSignature")) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        CFRelease(error);
+        goto on_error;
+    }
+
+    struct aws_byte_cursor signature_cur =
+        aws_byte_cursor_from_array(CFDataGetBytePtr(signature_ref), CFDataGetLength(signature_ref));
+
+    if (aws_byte_buf_append(out, &signature_cur)) {
+        aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+        goto on_error;
+    }
+
+    CFRelease(digest_ref);
+    CFRelease(signature_ref);
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+    CFRelease(digest_ref);
+
+    if (signature_ref != NULL) {
+        CFRelease(signature_ref);
+    }
+
+    return AWS_OP_ERR;
+}
+
+int s_rsa_verify(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signature_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_cursor signature) {
+    struct sec_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    if (key_pair_impl->pub_key_ref == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "RSA Key Pair is missing Public Key required for verify operation.");
+        return aws_raise_error(AWS_ERROR_CAL_MISSING_REQUIRED_KEY_COMPONENT);
+    }
+
+    SecKeyAlgorithm alg;
+    if (s_map_rsa_signing_algo_to_sec(algorithm, &alg)) {
+        return AWS_OP_ERR;
+    }
+
+    if (!SecKeyIsAlgorithmSupported(key_pair_impl->pub_key_ref, kSecKeyOperationTypeVerify, alg)) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Algo is not supported for this operation");
+        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    }
+
+    CFDataRef digest_ref = CFDataCreateWithBytesNoCopy(key_pair_impl->cf_allocator, digest.ptr, digest.len, kCFAllocatorNull);
+    CFDataRef signature_ref = CFDataCreateWithBytesNoCopy(NULL,
+        signature.ptr, signature.len, kCFAllocatorNull);
+    AWS_FATAL_ASSERT(digest_ref && signature_ref);
+
+    CFErrorRef error = NULL;
+    Boolean result = SecKeyVerifySignature(key_pair_impl->pub_key_ref, alg, digest_ref, signature_ref, &error);
+    
+    CFRelease(digest_ref);
+    CFRelease(signature_ref);
+    if (s_reinterpret_sec_error_as_crt(error, "SecKeyVerifySignature") || result == false) {
+        return aws_raise_error(AWS_ERROR_CAL_SIGNATURE_VALIDATION_FAILED);
+    }
```

Comment:
CRYPTO_OPERATION_FAILURE and VALIDATION_FAILURE are 2 separate errors

Also, forgot to free error
```suggestion
    if (s_reinterpret_sec_error_as_crt(error, "SecKeyVerifySignature")) {
        CFRelease(error);
        return AWS_OP_ERR;
    }
    
    if (result == false) {
        return aws_raise_error(AWS_ERROR_CAL_SIGNATURE_VALIDATION_FAILED);
    }
```

Reply:
updated to separate the errors out. this will not be consistent with ecc returns, but we can fix ecc to have better error handling later.